lib/blockstore/badger/badger.go

```diff
@@ -0,0 +1,360 @@
+
+	return b.DB.View(func(txn *badger.Txn) error {
+		switch item, err := txn.Get(k); err {
+		case nil:
+			return item.Value(fn)
+		case badger.ErrKeyNotFound:
+			return blockstore.ErrNotFound
+		default:
+			return fmt.Errorf("failed to view block from badger blockstore: %w", err)
+		}
+	})
+}
+
+func (b *Blockstore) Has(cid cid.Cid) (bool, error) {
+	if atomic.LoadInt64(&b.state) != stateOpen {
+		return false, ErrBlockstoreClosed
+	}
+
+	k, pooled := b.PooledPrefixedKey(cid)
+	if pooled {
+		defer pool.Put(k)
+	}
+
+	err := b.DB.View(func(txn *badger.Txn) error {
+		_, err := txn.Get(k)
+		return err
+	})
+
+	switch err {
+	case badger.ErrKeyNotFound:
+		return false, nil
+	case nil:
+		return true, nil
+	default:
+		return false, fmt.Errorf("failed to check if block exists in badger blockstore: %w", err)
+	}
+}
+
+func (b *Blockstore) Get(cid cid.Cid) (blocks.Block, error) {
+	if !cid.Defined() {
+		return nil, blockstore.ErrNotFound
+	}
+
+	if atomic.LoadInt64(&b.state) != stateOpen {
+		return nil, ErrBlockstoreClosed
+	}
+
+	k, pooled := b.PooledPrefixedKey(cid)
+	if pooled {
+		defer pool.Put(k)
+	}
+
+	var val []byte
+	err := b.DB.View(func(txn *badger.Txn) error {
+		switch item, err := txn.Get(k); err {
+		case nil:
+			val, err = item.ValueCopy(nil)
+			return err
+		case badger.ErrKeyNotFound:
+			return blockstore.ErrNotFound
+		default:
+			return fmt.Errorf("failed to get block from badger blockstore: %w", err)
+		}
+	})
+	if err != nil {
+		return nil, err
+	}
+	return blocks.NewBlockWithCid(val, cid)
+}
+
+func (b *Blockstore) GetSize(cid cid.Cid) (int, error) {
+	if atomic.LoadInt64(&b.state) != stateOpen {
+		return -1, ErrBlockstoreClosed
+	}
+
+	k, pooled := b.PooledPrefixedKey(cid)
+	if pooled {
+		defer pool.Put(k)
+	}
+
+	var size int
+	err := b.DB.View(func(txn *badger.Txn) error {
+		switch item, err := txn.Get(k); err {
+		case nil:
+			size = int(item.ValueSize())
+		case badger.ErrKeyNotFound:
+			return blockstore.ErrNotFound
+		default:
+			return fmt.Errorf("failed to get block size from badger blockstore: %w", err)
+		}
+		return nil
+	})
+	if err != nil {
+		size = -1
+	}
+	return size, err
+}
+
+func (b *Blockstore) Put(block blocks.Block) error {
+	if atomic.LoadInt64(&b.state) != stateOpen {
+		return ErrBlockstoreClosed
+	}
+
+	k, pooled := b.PooledPrefixedKey(block.Cid())
+	if pooled {
+		defer pool.Put(k)
+	}
+
+	err := b.DB.Update(func(txn *badger.Txn) error {
+		return txn.Set(k, block.RawData())
+	})
+	if err != nil {
+		err = fmt.Errorf("failed to put block in badger blockstore: %w", err)
+	}
+	return err
+}
+
+func (b *Blockstore) PutMany(blocks []blocks.Block) error {
+	if atomic.LoadInt64(&b.state) != stateOpen {
+		return ErrBlockstoreClosed
+	}
+
+	batch := b.DB.NewWriteBatch()
+	defer batch.Cancel()
+
+	// toReturn tracks the byte slices to return to the pool, if we're using key
+	// prefixing. we can't return each slice to the pool after each Set, because
+	// badger holds on to the slice.
```

Comment:
Unfortunately, this means that we return key byte slices to the pool once the full write batch is complete. Badger doesn't support callbacks that notify us interim, so it's all or nothing.